src/services/ghost/GhostStreamingParser.ts

```diff
@@ -0,0 +1,510 @@
+		if (incompleteChangeMatch || incompleteSearchMatch || incompleteReplaceMatch || incompleteCDataMatch) {
+			return false
+		}
+
+		// If the buffer is empty or only whitespace, consider it complete
+		if (trimmedBuffer.length === 0) {
+			return true
+		}
+
+		// If we have at least one complete change and no incomplete tags, likely complete
+		return this.completedChanges.length > 0
+	}
+
+	/**
+	 * Generate suggestions from completed changes
+	 */
+	private generateSuggestions(changes: ParsedChange[]): GhostSuggestionsState {
+		const suggestions = new GhostSuggestionsState()
+
+		if (!this.context?.document || changes.length === 0) {
+			return suggestions
+		}
+
+		const document = this.context.document
+		const currentContent = document.getText()
+
+		// Add cursor marker to document content if it's not already there
+		// This ensures that when LLM searches for <<<AUTOCOMPLETE_HERE>>>, it can find it
+		let modifiedContent = currentContent
+		const needsCursorMarker =
+			changes.some((change) => change.search.includes(CURSOR_MARKER)) && !currentContent.includes(CURSOR_MARKER)
+		if (needsCursorMarker && this.context.range) {
+			// Add cursor marker at the specified range position
+			const cursorOffset = document.offsetAt(this.context.range.start)
+			modifiedContent =
+				currentContent.substring(0, cursorOffset) + CURSOR_MARKER + currentContent.substring(cursorOffset)
+		}
+
+		// Process changes: preserve search content as-is, clean replace content for application
+		const filteredChanges = changes.map((change) => ({
+			search: change.search, // Keep cursor markers for matching against document
+			replace: this.removeCursorMarker(change.replace), // Clean for content application
+			cursorPosition: change.cursorPosition,
+		}))
+
+		// Apply changes in reverse order to maintain line numbers
+		const appliedChanges: Array<{
+			searchContent: string
+			replaceContent: string
+			startIndex: number
+			endIndex: number
+			cursorPosition?: number
+		}> = []
+
+		for (const change of filteredChanges) {
+			let searchIndex = this.findBestMatch(modifiedContent, change.search)
+
+			if (searchIndex !== -1) {
+				// Check for overlapping changes before applying
+				const endIndex = searchIndex + change.search.length
+				const hasOverlap = appliedChanges.some((existingChange) => {
+					// Check if ranges overlap
+					const existingStart = existingChange.startIndex
+					const existingEnd = existingChange.endIndex
+					return searchIndex < existingEnd && endIndex > existingStart
+				})
+
+				if (hasOverlap) {
+					console.warn("Skipping overlapping change:", change.search.substring(0, 50))
+					continue // Skip this change to avoid duplicates
+				}
+
+				// Handle the case where search pattern ends with newline but we need to preserve additional whitespace
+				let adjustedReplaceContent = change.replace
+
+				// If the search pattern ends with a newline, check if there are additional empty lines after it
```

Comment:
Could you add one more line here explaining _why_ we need to preserve it here? I assume if we don't, we accidentally remove it when we apply the suggestion?